File: server/models/match.ts
import { Schema, model } from 'mongoose';
import { IMatch } from '../types';

const matchSchema = new Schema<IMatch>({
  white: { type: Schema.Types.ObjectId, ref: "User" },
  black: { type: Schema.Types.ObjectId, ref: "User" },
  table: Number,
  result: {
    type: String,
    enum: ["IN_PROGRESS", "WHITE_VICTORY", "BLACK_VICTORY", "DRAW"]
  }
});

export default model<IMatch>("Match", matchSchema);

File: api/models/match.go
package models

import "go.mongodb.org/mongo-driver/bson/primitive"

type MatchResult string

const (
	InProgress  MatchResult = "IN_PROGRESS"
	WhiteWinner             = "WHITE_VICTORY"
	BlackWinner             = "BLACK_VICTORY"
	Draw                    = "DRAW"
)

type Match struct {
	Id        primitive.ObjectID `json:"_id" bson:"_id"`
	WhiteUser primitive.ObjectID `json:"white" bson:"white"`
	BlackUser primitive.ObjectID `json:"black" bson:"black"`
	Result    MatchResult        `json:"result" bson:"result"`
	Table     int                `json:"table" bson:"table"`
}

type PopulatedMatch struct {
	Id        primitive.ObjectID `json:"_id" bson:"_id"`
	WhiteUser User               `json:"white" bson:"white"`
	BlackUser User               `json:"black" bson:"black"`
	Result    MatchResult        `json:"result" bson:"result"`
	Table     int                `json:"table" bson:"table"`
}
